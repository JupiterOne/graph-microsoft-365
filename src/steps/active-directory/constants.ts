// Step IDs
export const STEP_ACCOUNT = 'account';
export const STEP_GROUPS = 'groups';
export const STEP_GROUP_MEMBERS = 'group-members';
export const STEP_USERS = 'users';

// Graph objects
export const ACCOUNT_ENTITY_TYPE = 'microsoft_365_account';
export const ACCOUNT_ENTITY_CLASS = 'Account';

export const GROUP_ENTITY_TYPE = 'microsoft_365_user_group';
export const GROUP_ENTITY_CLASS = 'UserGroup';

export const USER_ENTITY_TYPE = 'microsoft_365_user';
export const USER_ENTITY_CLASS = 'User';

/**
 * The entity type used for members of groups which are not one of the ingested
 * directory objects.
 */
export const GROUP_MEMBER_ENTITY_TYPE = 'microsoft_365_group_member';

/**
 * The entity class used for members of groups which are not one of the ingested
 * directory objects.
 */
export const GROUP_MEMBER_ENTITY_CLASS = 'User';

export const ACCOUNT_GROUP_RELATIONSHIP_TYPE =
  'microsoft_365_account_has_group';
export const ACCOUNT_USER_RELATIONSHIP_TYPE = 'microsoft_365_account_has_user';
export const GROUP_MEMBER_RELATIONSHIP_TYPE = 'microsoft_365_group_has_member';
