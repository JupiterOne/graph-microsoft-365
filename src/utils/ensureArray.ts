import { isArray } from 'lodash';

/**
 * Convert a string or string array value to a string array
 */
export function ensureArray<T>(val: T | T[]): T[] {
  return isArray(val) ? (val as T[]) : [val as T];
}
