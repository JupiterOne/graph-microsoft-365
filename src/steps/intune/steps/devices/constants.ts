import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';
import { entities as activeDirectoryEntities } from '../../../active-directory';

export const steps: { [k: string]: string } = {
  FETCH_DEVICES: 'managed-devices',
};

export const entities: { [k: string]: StepEntityMetadata } = {
  DEVICE: {
    resourceName: 'Managed Device',
    _type: 'intune_managed_device',
    _class: ['Host', 'Device'],
  },
};

export const relationships: { [k: string]: StepRelationshipMetadata } = {
  USER_HAS_DEVICE: {
    _type: 'microsoft_365_user_has_intune_managed_device',
    sourceType: activeDirectoryEntities.USER._type,
    _class: RelationshipClass.HAS,
    targetType: entities.DEVICE._type,
  },
};
