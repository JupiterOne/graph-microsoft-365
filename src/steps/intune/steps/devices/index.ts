import {
  IntegrationStepExecutionContext,
  Step,
} from '@jupiterone/integration-sdk-core';
import { IntegrationConfig, IntegrationStepContext } from '../../../../types';
import { steps as activeDirectorySteps } from '../../../active-directory';
import { DeviceManagementIntuneClient } from '../../clients/deviceManagementIntuneClient';
import { relationships, entities, steps } from './constants';
import {
  createManagedDeviceEntity,
  createUserDeviceDirectRelationship,
  createUserDeviceMappedRelationship,
} from './converters';

export async function fetchDevices(
  executionContext: IntegrationStepContext,
): Promise<void> {
  const { logger, instance, jobState } = executionContext;
  const intuneClient = new DeviceManagementIntuneClient(
    logger,
    instance.config,
  );

  await intuneClient.iterateManagedDevices(async (device) => {
    const deviceEntity = createManagedDeviceEntity(device);
    await jobState.addEntity(deviceEntity);
    const userEntity = await jobState.findEntity(device.userId as string);
    const userDeviceRelationship = userEntity
      ? createUserDeviceDirectRelationship(deviceEntity, userEntity)
      : createUserDeviceMappedRelationship(
          deviceEntity,
          device.userId as string,
          device.emailAddress as string,
        );
    if (userDeviceRelationship) {
      await jobState.addRelationship(userDeviceRelationship);
    }
  });
}

export const deviceSteps: Step<
  IntegrationStepExecutionContext<IntegrationConfig>
>[] = [
  {
    id: steps.FETCH_DEVICES,
    name: 'Managed Devices',
    entities: [entities.DEVICE],
    relationships: [relationships.USER_HAS_DEVICE],
    dependsOn: [activeDirectorySteps.FETCH_USERS],
    executionHandler: fetchDevices,
  },
];
