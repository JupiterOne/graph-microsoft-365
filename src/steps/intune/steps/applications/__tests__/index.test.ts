import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { setupAzureRecording } from '../../../../../../test/recording';
import { config } from '../../../../../../test/config';
import { fetchDevices } from '../../devices';
import { entities, relationships } from '../../../constants';
import { isEqual } from 'lodash';
import { ensureArray } from '../../../../../../test/ensureArray';
import { fetchManagedApplications } from '..';

let recording: Recording;

afterEach(async () => {
  if (recording) {
    await recording.stop();
  }
});

describe('fetchManagedApplications', () => {
  test('should make entities and relationships correctly', async () => {
    recording = setupAzureRecording({
      directory: __dirname,
      name: 'fetchManagedApplications',
    });
    const context = createMockStepExecutionContext({ instanceConfig: config });

    await fetchDevices(context);
    await fetchManagedApplications(context);

    const managedApplicationEntities = context.jobState.collectedEntities.filter(
      (e) =>
        isEqual(e._class, ensureArray(entities.MANAGED_APPLICATION._class)),
    );
    const deviceApplicationRelationships = context.jobState.collectedRelationships.filter(
      (r) =>
        isEqual(
          r._type,
          relationships.DEVICE_ASSIGNED_MANAGED_APPLICATION._type,
        ),
    );

    // Check that we have Managed Applications
    expect(managedApplicationEntities.length).toBeGreaterThan(0);
    expect(managedApplicationEntities).toMatchGraphObjectSchema({
      _class: entities.MANAGED_APPLICATION._class,
    });
    expect(managedApplicationEntities).toMatchSnapshot(
      'managedApplicationEntities',
    );

    // Check that we have DEVICE_ASSIGNED_MANAGED_APPLICATION relationships
    expect(deviceApplicationRelationships.length).toBeGreaterThan(0);
    expect(deviceApplicationRelationships).toMatchDirectRelationshipSchema({});
    expect(deviceApplicationRelationships).toMatchSnapshot(
      'deviceApplicationRelationships',
    );
  });
});
