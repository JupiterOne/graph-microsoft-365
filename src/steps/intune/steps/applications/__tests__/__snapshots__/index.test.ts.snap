// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchManagedApplications should make entities and relationships correctly: deviceApplicationRelationships 1`] = `
Array [
  Object {
    "_class": "ASSIGNED",
    "_fromEntityKey": "683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_key": "683dbff1-c8ff-4996-91ae-85484de46cfc|assigned|50124056-8cca-4cf5-9f98-8e63bede0b6e",
    "_toEntityKey": "50124056-8cca-4cf5-9f98-8e63bede0b6e",
    "_type": "intune_managed_device_assigned_application",
    "displayName": "ASSIGNED",
    "errorCode": 0,
    "installState": "installed",
    "installStateDetail": "noAdditionalDetails",
  },
  Object {
    "_class": "ASSIGNED",
    "_fromEntityKey": "1f75329e-51fc-451c-bd40-c457337641a4",
    "_key": "1f75329e-51fc-451c-bd40-c457337641a4|assigned|55c4e44f-7e06-4586-8f4c-b1aebf301d32",
    "_toEntityKey": "55c4e44f-7e06-4586-8f4c-b1aebf301d32",
    "_type": "intune_managed_device_assigned_application",
    "displayName": "ASSIGNED",
    "errorCode": 0,
    "installState": "installed",
    "installStateDetail": "noAdditionalDetails",
  },
]
`;

exports[`fetchManagedApplications should make entities and relationships correctly: managedApplicationEntities 1`] = `
Array [
  Object {
    "COTS": false,
    "_class": Array [
      "Application",
    ],
    "_key": "cfbeb753-49e3-4ed5-9743-f857eb74df2a",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "@odata.type": "#microsoft.graph.androidLobApp",
          "committedContentVersion": null,
          "createdDateTime": "2021-03-01T22:59:11Z",
          "dependentAppCount": 0,
          "description": "com.android.vending.apk",
          "developer": "developer me",
          "displayName": "com.android.vending.apk",
          "fileName": "com.android.vending.apk",
          "id": "cfbeb753-49e3-4ed5-9743-f857eb74df2a",
          "identityName": "com.android.vending",
          "identityVersion": "705",
          "informationUrl": null,
          "isAssigned": false,
          "isFeatured": false,
          "largeIcon": null,
          "lastModifiedDateTime": "2021-03-01T22:59:11Z",
          "minimumSupportedOperatingSystem": Object {
            "v10_0": false,
            "v11_0": false,
            "v4_0": true,
            "v4_0_3": false,
            "v4_1": false,
            "v4_2": false,
            "v4_3": false,
            "v4_4": false,
            "v5_0": false,
            "v5_1": false,
            "v6_0": false,
            "v7_0": false,
            "v7_1": false,
            "v8_0": false,
            "v8_1": false,
            "v9_0": false,
          },
          "notes": "notes",
          "owner": "owner me",
          "packageId": "com.android.vending",
          "privacyInformationUrl": "",
          "publisher": "Noodle's n No Company",
          "publishingState": "notPublished",
          "roleScopeTagIds": Array [],
          "size": 0,
          "supersededAppCount": 0,
          "supersedingAppCount": 0,
          "uploadState": 0,
          "versionCode": "705",
          "versionName": "0.7.5",
        },
      },
    ],
    "_type": "intune_managed_application",
    "committedContentVersion": null,
    "createdOn": 1614639551000,
    "description": "com.android.vending.apk",
    "developer": "developer me",
    "displayName": "com.android.vending.apk",
    "external": false,
    "id": "cfbeb753-49e3-4ed5-9743-f857eb74df2a",
    "informationURL": null,
    "isFeatured": false,
    "isPublished": false,
    "lastUpdatedOn": 1614639551000,
    "mobile": true,
    "name": "com.android.vending.apk",
    "notes": Array [
      "notes",
    ],
    "owner": "owner me",
    "packageId": "com.android.vending",
    "privacyInformationURL": "",
    "productionURL": undefined,
    "publisher": "Noodle's n No Company",
    "version": "705",
  },
  Object {
    "COTS": true,
    "_class": Array [
      "Application",
    ],
    "_key": "50124056-8cca-4cf5-9f98-8e63bede0b6e",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "@odata.type": "#microsoft.graph.webApp",
          "appUrl": "https://www.google.com",
          "createdDateTime": "2021-02-17T18:04:47Z",
          "dependentAppCount": 0,
          "description": "just google.com",
          "developer": "Google",
          "displayName": "google.com",
          "id": "50124056-8cca-4cf5-9f98-8e63bede0b6e",
          "informationUrl": null,
          "isAssigned": true,
          "isFeatured": true,
          "largeIcon": null,
          "lastModifiedDateTime": "2021-02-17T18:04:47Z",
          "notes": "",
          "owner": "",
          "privacyInformationUrl": "",
          "publisher": "Google",
          "publishingState": "published",
          "roleScopeTagIds": Array [],
          "supersededAppCount": 0,
          "supersedingAppCount": 0,
          "uploadState": 1,
          "useManagedBrowser": false,
        },
      },
    ],
    "_type": "intune_managed_application",
    "committedContentVersion": undefined,
    "createdOn": 1613585087000,
    "description": "just google.com",
    "developer": "Google",
    "displayName": "google.com",
    "external": true,
    "id": "50124056-8cca-4cf5-9f98-8e63bede0b6e",
    "informationURL": null,
    "isFeatured": true,
    "isPublished": true,
    "lastUpdatedOn": 1613585087000,
    "mobile": false,
    "name": "google.com",
    "notes": Array [],
    "owner": undefined,
    "packageId": undefined,
    "privacyInformationURL": "",
    "productionURL": "https://www.google.com",
    "publisher": "Google",
    "version": undefined,
  },
  Object {
    "COTS": true,
    "_class": Array [
      "Application",
    ],
    "_key": "813a7159-4c66-4286-82f3-51b39aaefcd9",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "@odata.type": "#microsoft.graph.androidManagedStoreApp",
          "appIdentifier": "com.microsoft.windowsintune.companyportal",
          "appStoreUrl": "https://play.google.com/store/apps/details?id=com.microsoft.windowsintune.companyportal&hl=en-US",
          "appTracks": Array [],
          "createdDateTime": "2021-02-17T17:02:31Z",
          "dependentAppCount": 0,
          "description": "Install company apps & remotely manage your devices enrolled to Microsoft Intune",
          "developer": null,
          "displayName": "Intune Company Portal",
          "id": "813a7159-4c66-4286-82f3-51b39aaefcd9",
          "informationUrl": "https://play.google.com/store/apps/details?id=com.microsoft.windowsintune.companyportal&hl=en-US",
          "isAssigned": true,
          "isFeatured": false,
          "isPrivate": false,
          "isSystemApp": false,
          "largeIcon": null,
          "lastModifiedDateTime": "2021-02-17T17:02:31Z",
          "notes": null,
          "owner": null,
          "packageId": "com.microsoft.windowsintune.companyportal",
          "privacyInformationUrl": null,
          "publisher": "Microsoft Corporation",
          "publishingState": "published",
          "roleScopeTagIds": Array [],
          "supersededAppCount": 0,
          "supersedingAppCount": 0,
          "supportsOemConfig": false,
          "totalLicenseCount": 0,
          "uploadState": 1,
          "usedLicenseCount": 0,
        },
      },
    ],
    "_type": "intune_managed_application",
    "committedContentVersion": undefined,
    "createdOn": 1613581351000,
    "description": "Install company apps & remotely manage your devices enrolled to Microsoft Intune",
    "developer": null,
    "displayName": "Intune Company Portal",
    "external": true,
    "id": "813a7159-4c66-4286-82f3-51b39aaefcd9",
    "informationURL": "https://play.google.com/store/apps/details?id=com.microsoft.windowsintune.companyportal&hl=en-US",
    "isFeatured": false,
    "isPublished": true,
    "lastUpdatedOn": 1613581351000,
    "mobile": true,
    "name": "Intune Company Portal",
    "notes": Array [],
    "owner": undefined,
    "packageId": "com.microsoft.windowsintune.companyportal",
    "privacyInformationURL": null,
    "productionURL": "https://play.google.com/store/apps/details?id=com.microsoft.windowsintune.companyportal&hl=en-US",
    "publisher": "Microsoft Corporation",
    "version": undefined,
  },
  Object {
    "COTS": true,
    "_class": Array [
      "Application",
    ],
    "_key": "ccc52be9-a284-4f17-b669-f21a11bb4668",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "@odata.type": "#microsoft.graph.androidManagedStoreApp",
          "appIdentifier": "com.microsoft.launcher.enterprise",
          "appStoreUrl": "https://play.google.com/store/apps/details?id=com.microsoft.launcher.enterprise&hl=en-US",
          "appTracks": Array [],
          "createdDateTime": "2021-02-17T17:02:31Z",
          "dependentAppCount": 0,
          "description": "Managed Home Screen",
          "developer": null,
          "displayName": "Managed Home Screen",
          "id": "ccc52be9-a284-4f17-b669-f21a11bb4668",
          "informationUrl": "https://play.google.com/store/apps/details?id=com.microsoft.launcher.enterprise&hl=en-US",
          "isAssigned": false,
          "isFeatured": false,
          "isPrivate": false,
          "isSystemApp": false,
          "largeIcon": null,
          "lastModifiedDateTime": "2021-02-17T17:02:31Z",
          "notes": null,
          "owner": null,
          "packageId": "com.microsoft.launcher.enterprise",
          "privacyInformationUrl": null,
          "publisher": "Microsoft Corporation",
          "publishingState": "published",
          "roleScopeTagIds": Array [],
          "supersededAppCount": 0,
          "supersedingAppCount": 0,
          "supportsOemConfig": false,
          "totalLicenseCount": 0,
          "uploadState": 1,
          "usedLicenseCount": 0,
        },
      },
    ],
    "_type": "intune_managed_application",
    "committedContentVersion": undefined,
    "createdOn": 1613581351000,
    "description": "Managed Home Screen",
    "developer": null,
    "displayName": "Managed Home Screen",
    "external": true,
    "id": "ccc52be9-a284-4f17-b669-f21a11bb4668",
    "informationURL": "https://play.google.com/store/apps/details?id=com.microsoft.launcher.enterprise&hl=en-US",
    "isFeatured": false,
    "isPublished": true,
    "lastUpdatedOn": 1613581351000,
    "mobile": true,
    "name": "Managed Home Screen",
    "notes": Array [],
    "owner": undefined,
    "packageId": "com.microsoft.launcher.enterprise",
    "privacyInformationURL": null,
    "productionURL": "https://play.google.com/store/apps/details?id=com.microsoft.launcher.enterprise&hl=en-US",
    "publisher": "Microsoft Corporation",
    "version": undefined,
  },
  Object {
    "COTS": true,
    "_class": Array [
      "Application",
    ],
    "_key": "55c4e44f-7e06-4586-8f4c-b1aebf301d32",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "@odata.type": "#microsoft.graph.macOSOfficeSuiteApp",
          "createdDateTime": "2021-02-17T17:45:22.1995185Z",
          "dependentAppCount": 0,
          "description": "Microsoft 365 Apps for macOS",
          "developer": "Microsoft",
          "displayName": "Microsoft 365 Apps for macOS",
          "id": "55c4e44f-7e06-4586-8f4c-b1aebf301d32",
          "informationUrl": "https://products.office.com/en-us/explore-office-for-home",
          "isAssigned": true,
          "isFeatured": true,
          "largeIcon": null,
          "lastModifiedDateTime": "2021-02-18T20:32:36.7100976Z",
          "notes": "",
          "owner": "Microsoft",
          "privacyInformationUrl": "https://privacy.microsoft.com/en-US/privacystatement",
          "publisher": "Microsoft",
          "publishingState": "published",
          "roleScopeTagIds": Array [],
          "supersededAppCount": 0,
          "supersedingAppCount": 0,
          "uploadState": 1,
        },
      },
    ],
    "_type": "intune_managed_application",
    "committedContentVersion": undefined,
    "createdOn": 1613583922199,
    "description": "Microsoft 365 Apps for macOS",
    "developer": "Microsoft",
    "displayName": "Microsoft 365 Apps for macOS",
    "external": true,
    "id": "55c4e44f-7e06-4586-8f4c-b1aebf301d32",
    "informationURL": "https://products.office.com/en-us/explore-office-for-home",
    "isFeatured": true,
    "isPublished": true,
    "lastUpdatedOn": 1613680356710,
    "mobile": false,
    "name": "Microsoft 365 Apps for macOS",
    "notes": Array [],
    "owner": "Microsoft",
    "packageId": undefined,
    "privacyInformationURL": "https://privacy.microsoft.com/en-US/privacystatement",
    "productionURL": undefined,
    "publisher": "Microsoft",
    "version": undefined,
  },
  Object {
    "COTS": true,
    "_class": Array [
      "Application",
    ],
    "_key": "d88ed6fd-24da-406e-aba6-3e2c0c7a7e1d",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "@odata.type": "#microsoft.graph.officeSuiteApp",
          "autoAcceptEula": false,
          "createdDateTime": "2021-02-16T22:24:49Z",
          "dependentAppCount": 0,
          "description": "Microsoft 365 Apps for Windows 10",
          "developer": "Microsoft",
          "displayName": "Microsoft 365 Apps for Windows 10",
          "excludedApps": Object {
            "access": false,
            "bing": false,
            "excel": false,
            "groove": true,
            "infoPath": true,
            "lync": false,
            "oneDrive": false,
            "oneNote": false,
            "outlook": false,
            "powerPoint": false,
            "publisher": false,
            "sharePointDesigner": true,
            "teams": false,
            "visio": false,
            "word": false,
          },
          "id": "d88ed6fd-24da-406e-aba6-3e2c0c7a7e1d",
          "informationUrl": "https://products.office.com/en-us/explore-office-for-home",
          "installProgressDisplayLevel": "none",
          "isAssigned": true,
          "isFeatured": false,
          "largeIcon": null,
          "lastModifiedDateTime": "2021-02-16T22:24:49Z",
          "localesToInstall": Array [],
          "notes": "",
          "officeConfigurationXml": null,
          "officePlatformArchitecture": "x64",
          "owner": "Microsoft",
          "privacyInformationUrl": "https://privacy.microsoft.com/en-US/privacystatement",
          "productIds": Array [
            "o365ProPlusRetail",
          ],
          "publisher": "Microsoft",
          "publishingState": "published",
          "roleScopeTagIds": Array [],
          "shouldUninstallOlderVersionsOfOffice": true,
          "supersededAppCount": 0,
          "supersedingAppCount": 0,
          "targetVersion": "",
          "updateChannel": "current",
          "updateVersion": "",
          "uploadState": 1,
          "useSharedComputerActivation": true,
        },
      },
    ],
    "_type": "intune_managed_application",
    "committedContentVersion": undefined,
    "createdOn": 1613514289000,
    "description": "Microsoft 365 Apps for Windows 10",
    "developer": "Microsoft",
    "displayName": "Microsoft 365 Apps for Windows 10",
    "external": true,
    "id": "d88ed6fd-24da-406e-aba6-3e2c0c7a7e1d",
    "informationURL": "https://products.office.com/en-us/explore-office-for-home",
    "isFeatured": false,
    "isPublished": true,
    "lastUpdatedOn": 1613514289000,
    "mobile": false,
    "name": "Microsoft 365 Apps for Windows 10",
    "notes": Array [],
    "owner": "Microsoft",
    "packageId": undefined,
    "privacyInformationURL": "https://privacy.microsoft.com/en-US/privacystatement",
    "productionURL": undefined,
    "publisher": "Microsoft",
    "version": undefined,
  },
  Object {
    "COTS": true,
    "_class": Array [
      "Application",
    ],
    "_key": "f9872827-46cb-4d02-80d8-063fa5a9588b",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "@odata.type": "#microsoft.graph.androidManagedStoreApp",
          "appIdentifier": "com.azure.authenticator",
          "appStoreUrl": "https://play.google.com/store/apps/details?id=com.azure.authenticator&hl=en-US",
          "appTracks": Array [],
          "createdDateTime": "2021-02-17T17:02:31Z",
          "dependentAppCount": 0,
          "description": "No more passwords, easier sign-in, and greater protection for your accounts.",
          "developer": null,
          "displayName": "Microsoft Authenticator",
          "id": "f9872827-46cb-4d02-80d8-063fa5a9588b",
          "informationUrl": "https://play.google.com/store/apps/details?id=com.azure.authenticator&hl=en-US",
          "isAssigned": false,
          "isFeatured": false,
          "isPrivate": false,
          "isSystemApp": false,
          "largeIcon": null,
          "lastModifiedDateTime": "2021-02-17T17:02:31Z",
          "notes": null,
          "owner": null,
          "packageId": "com.azure.authenticator",
          "privacyInformationUrl": null,
          "publisher": "Microsoft Corporation",
          "publishingState": "published",
          "roleScopeTagIds": Array [],
          "supersededAppCount": 0,
          "supersedingAppCount": 0,
          "supportsOemConfig": false,
          "totalLicenseCount": 0,
          "uploadState": 1,
          "usedLicenseCount": 0,
        },
      },
    ],
    "_type": "intune_managed_application",
    "committedContentVersion": undefined,
    "createdOn": 1613581351000,
    "description": "No more passwords, easier sign-in, and greater protection for your accounts.",
    "developer": null,
    "displayName": "Microsoft Authenticator",
    "external": true,
    "id": "f9872827-46cb-4d02-80d8-063fa5a9588b",
    "informationURL": "https://play.google.com/store/apps/details?id=com.azure.authenticator&hl=en-US",
    "isFeatured": false,
    "isPublished": true,
    "lastUpdatedOn": 1613581351000,
    "mobile": true,
    "name": "Microsoft Authenticator",
    "notes": Array [],
    "owner": undefined,
    "packageId": "com.azure.authenticator",
    "privacyInformationURL": null,
    "productionURL": "https://play.google.com/store/apps/details?id=com.azure.authenticator&hl=en-US",
    "publisher": "Microsoft Corporation",
    "version": undefined,
  },
  Object {
    "COTS": true,
    "_class": Array [
      "Application",
    ],
    "_key": "d2c6709c-bec1-4e94-bcbc-0c7835ff19d9",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "@odata.type": "#microsoft.graph.androidManagedStoreApp",
          "appIdentifier": "com.microsoft.intune",
          "appStoreUrl": "https://play.google.com/store/apps/details?id=com.microsoft.intune&hl=en-US",
          "appTracks": Array [],
          "createdDateTime": "2021-02-17T17:02:31Z",
          "dependentAppCount": 0,
          "description": "Get access to your organization’s resources and keep them secure.",
          "developer": null,
          "displayName": "Microsoft Intune",
          "id": "d2c6709c-bec1-4e94-bcbc-0c7835ff19d9",
          "informationUrl": "https://play.google.com/store/apps/details?id=com.microsoft.intune&hl=en-US",
          "isAssigned": false,
          "isFeatured": false,
          "isPrivate": false,
          "isSystemApp": false,
          "largeIcon": null,
          "lastModifiedDateTime": "2021-02-17T17:02:31Z",
          "notes": null,
          "owner": null,
          "packageId": "com.microsoft.intune",
          "privacyInformationUrl": null,
          "publisher": "Microsoft Corporation",
          "publishingState": "published",
          "roleScopeTagIds": Array [],
          "supersededAppCount": 0,
          "supersedingAppCount": 0,
          "supportsOemConfig": false,
          "totalLicenseCount": 0,
          "uploadState": 1,
          "usedLicenseCount": 0,
        },
      },
    ],
    "_type": "intune_managed_application",
    "committedContentVersion": undefined,
    "createdOn": 1613581351000,
    "description": "Get access to your organization’s resources and keep them secure.",
    "developer": null,
    "displayName": "Microsoft Intune",
    "external": true,
    "id": "d2c6709c-bec1-4e94-bcbc-0c7835ff19d9",
    "informationURL": "https://play.google.com/store/apps/details?id=com.microsoft.intune&hl=en-US",
    "isFeatured": false,
    "isPublished": true,
    "lastUpdatedOn": 1613581351000,
    "mobile": true,
    "name": "Microsoft Intune",
    "notes": Array [],
    "owner": undefined,
    "packageId": "com.microsoft.intune",
    "privacyInformationURL": null,
    "productionURL": "https://play.google.com/store/apps/details?id=com.microsoft.intune&hl=en-US",
    "publisher": "Microsoft Corporation",
    "version": undefined,
  },
  Object {
    "COTS": true,
    "_class": Array [
      "Application",
    ],
    "_key": "152c83ee-8d4d-4fd4-8cf3-62f7f6689ca0",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "@odata.type": "#microsoft.graph.iosStoreApp",
          "appStoreUrl": "https://apps.apple.com/us/app/microsoft-word/id586447913?uo=4",
          "applicableDeviceType": Object {
            "iPad": true,
            "iPhoneAndIPod": true,
          },
          "bundleId": "com.microsoft.Office.Word",
          "createdDateTime": "2021-02-18T23:18:44.6113093Z",
          "dependentAppCount": 0,
          "description": "The trusted Word app lets you create, edit, view, and share your files with others quickly and easily. Send, view and edit Office docs attached to emails from your phone with this powerful word processing app from Microsoft.

With Word, your office moves with you. Whether you’re a blogger, writer, journalist, columnist, student, or a project manager working on documentation, Word is as handy as you want it to be.

Make impactful text documents, scripts, blogs, write-ups, or resumes. Customize your document, letter, resume, or notes your way with robust tools that enable you to accomplish your best writing with the best format options and templates.

Get the complete Microsoft Office experience when you sign in with your Microsoft 365 subscription. Create and edit documents, collaborate and share with ease when you download Microsoft Word today.

Create Documents
• Jump-start your projects, assignments, letters, blogs, scripts, notes, write-ups, or resume with beautifully designed modern templates.
• Use rich formatting and layout options to note down your ideas and express them in writing.
• Document formats and layouts stay pristine and look great on any device.

Read, Write, & Edit Documents
• Reading view lets you read lengthy docs, letters, scripts, and more on your device with ease.
• Use familiar writing and editing tools as you’d find on a desktop word processor.

Collaborate Anywhere
• Share thoughts by commenting in docs right next to the text you’re discussing.
• Everyone can collaborate and stay on top of changes to the text, layout, and formatting.
• As you and your team make changes, you can revert to view earlier drafts with improved version history in Word.

Simplified Sharing
• Share files with a few taps to quickly invite others to edit or view your documents.
• Easily manage permissions and see who’s working in a document.
• Copy the content of your Word files directly into the body of an email message with its format intact or attach your docs to an email and make sharing easier.

Download Microsoft Word now and get the best way to create, share and edit documents from your iPhone or iPad.

To create or edit documents, sign in with a free Microsoft account on devices with a screen size smaller than 10.1 inches.
Unlock the full Microsoft Office experience with a qualifying Microsoft 365 subscription for your phone, tablet, PC, and Mac.

Monthly Microsoft 365 subscriptions purchased from the app will be charged to your App Store account and will automatically renew within 24 hours prior to the end of the current subscription period unless auto-renewal is disabled beforehand. You can manage your subscriptions in your App Store account settings.

This app is provided by either Microsoft or a third-party app publisher and is subject to a separate privacy statement and terms and conditions. Data provided through the use of this store and this app may be accessible to Microsoft or the third-party app publisher, as applicable, and transferred to, stored, and processed in the United States or any other country where Microsoft or the app publisher and their affiliates or service providers maintain facilities.

Please refer to the Microsoft Software License Terms for Microsoft 365. See “License Agreement” link under Information. By installing the app, you agree to these terms and conditions.",
          "developer": "",
          "displayName": "Microsoft Word",
          "id": "152c83ee-8d4d-4fd4-8cf3-62f7f6689ca0",
          "informationUrl": null,
          "isAssigned": true,
          "isFeatured": false,
          "largeIcon": null,
          "lastModifiedDateTime": "2021-02-18T23:18:44.6113093Z",
          "minimumSupportedOperatingSystem": Object {
            "v10_0": false,
            "v11_0": false,
            "v12_0": false,
            "v13_0": false,
            "v14_0": false,
            "v8_0": true,
            "v9_0": false,
          },
          "notes": "",
          "owner": "",
          "privacyInformationUrl": "",
          "publisher": "Microsoft Corporation",
          "publishingState": "published",
          "roleScopeTagIds": Array [],
          "supersededAppCount": 0,
          "supersedingAppCount": 0,
          "uploadState": 1,
        },
      },
    ],
    "_type": "intune_managed_application",
    "committedContentVersion": undefined,
    "createdOn": 1613690324611,
    "description": "The trusted Word app lets you create, edit, view, and share your files with others quickly and easily. Send, view and edit Office docs attached to emails from your phone with this powerful word processing app from Microsoft.

With Word, your office moves with you. Whether you’re a blogger, writer, journalist, columnist, student, or a project manager working on documentation, Word is as handy as you want it to be.

Make impactful text documents, scripts, blogs, write-ups, or resumes. Customize your document, letter, resume, or notes your way with robust tools that enable you to accomplish your best writing with the best format options and templates.

Get the complete Microsoft Office experience when you sign in with your Microsoft 365 subscription. Create and edit documents, collaborate and share with ease when you download Microsoft Word today.

Create Documents
• Jump-start your projects, assignments, letters, blogs, scripts, notes, write-ups, or resume with beautifully designed modern templates.
• Use rich formatting and layout options to note down your ideas and express them in writing.
• Document formats and layouts stay pristine and look great on any device.

Read, Write, & Edit Documents
• Reading view lets you read lengthy docs, letters, scripts, and more on your device with ease.
• Use familiar writing and editing tools as you’d find on a desktop word processor.

Collaborate Anywhere
• Share thoughts by commenting in docs right next to the text you’re discussing.
• Everyone can collaborate and stay on top of changes to the text, layout, and formatting.
• As you and your team make changes, you can revert to view earlier drafts with improved version history in Word.

Simplified Sharing
• Share files with a few taps to quickly invite others to edit or view your documents.
• Easily manage permissions and see who’s working in a document.
• Copy the content of your Word files directly into the body of an email message with its format intact or attach your docs to an email and make sharing easier.

Download Microsoft Word now and get the best way to create, share and edit documents from your iPhone or iPad.

To create or edit documents, sign in with a free Microsoft account on devices with a screen size smaller than 10.1 inches.
Unlock the full Microsoft Office experience with a qualifying Microsoft 365 subscription for your phone, tablet, PC, and Mac.

Monthly Microsoft 365 subscriptions purchased from the app will be charged to your App Store account and will automatically renew within 24 hours prior to the end of the current subscription period unless auto-renewal is disabled beforehand. You can manage your subscriptions in your App Store account settings.

This app is provided by either Microsoft or a third-party app publisher and is subject to a separate privacy statement and terms and conditions. Data provided through the use of this store and this app may be accessible to Microsoft or the third-party app publisher, as applicable, and transferred to, stored, and processed in the United States or any other country where Microsoft or the app publisher and their affiliates or service providers maintain facilities.

Please refer to the Microsoft Software License Terms for Microsoft 365. See “License Agreement” link under Information. By installing the app, you agree to these terms and conditions.",
    "developer": "",
    "displayName": "Microsoft Word",
    "external": true,
    "id": "152c83ee-8d4d-4fd4-8cf3-62f7f6689ca0",
    "informationURL": null,
    "isFeatured": false,
    "isPublished": true,
    "lastUpdatedOn": 1613690324611,
    "mobile": true,
    "name": "Microsoft Word",
    "notes": Array [],
    "owner": undefined,
    "packageId": undefined,
    "privacyInformationURL": "",
    "productionURL": "https://apps.apple.com/us/app/microsoft-word/id586447913?uo=4",
    "publisher": "Microsoft Corporation",
    "version": undefined,
  },
]
`;
